biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java

```diff
@@ -77,16 +78,49 @@
 		validate();
 
 		Repository r;
-		if (this.indexFile.isFile())
-			r = readFile();
-		else {
+		if (!this.indexFile.isFile()) {
+			r = readRepository();
+			save(r);
+		} else if (!startsAnyResourceUrlWithPrefix(url)) {
 			r = readRepository();
 			save(r);
+		} else {
+			r = readFile();
 		}
 
 		bridge = new BridgeRepository(r);
 	}
 
+	private boolean startsAnyResourceUrlWithPrefix(URI prefix) throws Exception {
```

Comment:
This seems like a very convoluted way of deciding the URI changed. Since `name` is written into the saved index file

https://github.com/bndtools/bnd/blob/5d8500fd174eb7ec4228f383baba910d4d5573f8/biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java#L227

We should use a name based upon the URI 

https://github.com/bndtools/bnd/blob/5d8500fd174eb7ec4228f383baba910d4d5573f8/biz.aQute.bndlib/src/aQute/bnd/http/URLCache.java#L149-L151

as the name attribute of the repository element, we just need to see if the name based upon the current URI matches the name in the xml. But to do this, we will need to modify XMLResourceParser since it does not currently expose the name (and increment) attribute of the top level resource element.

With this change, the check for changed URI  become easy.

Reply:
I made the above suggested fix in PR #2165

Reply:
More elegant way. #PR 2165 is very helpful for that.